Handle nested options in multi-checkbox widget

markstory opened the pull request:
This makes converting from a multiple select box into a multiple-checkbox widget simple as it works out of the box. This also makes the behavior match the documentation. While this feature was documented, it did not have any tests in 2.x and thus wasn't preserved in the 3.x rewrite.

Refs #8385

markstory committed:
Handle nested options in multi-checkbox widget

This makes converting from a multiple select box into
a multiple-checkbox widget simple as it works out of the box. This also
makes the behavior match the documentation. While this feature was
documented, it did not have any tests in 2.x and thus wasn't preserved
in the 3.x rewrite.

Refs #8385

ndm2 commented on src/View/Widget/MultiCheckboxWidget.php:
```diff
@@ -113,10 +115,26 @@ public function render(array $data, ContextInterface $context)
+            if (is_string($key) && is_array($val) && !isset($val['text'], $val['value'])) {
+                $inputs = $this->_renderInputs(['options' => $val] + $data, $context);
+                $legend = $this->_templates->format('legend', ['text' => $key]);
+                $out[] = $this->_templates->format('fieldset', [
```
Would it be reasonable to use separate templates for text and container? While one may want to change the templates for multi-checkboxes, having this affect fieldsets and legends in general may be unwelcome.

Another option might be to set a (configurable) class for the fieldset, so that they can be easily targeted via CSS.

An additional template around the fieldsets so that everything is contained in a single element might be useful too.

markstory replied:
If this input widget was used with input() it would get a containing element from that method. When used in isolation, the developer could wrap the widget with raw HTML.

Are you suggesting adding a template like `multicheckboxcontainer` and `multicheckboxtitle` that default to a fieldset/legend?

ndm2 replied:
> If this input widget was used with input() it would get a containing element from that method

Looked at the wrong output sorry, input() generates a wrapper as expected.

> Are you suggesting adding a template like multicheckboxcontainer and multicheckboxtitle that default to a fieldset/legend?

Yep.

markstory replied:
Ok. I can add new templates. New templates should be backwards compatible as custom templates are layered on top of the defaults.

markstory committed:
Add discrete templates for multicheckbox elements.

This will let people modify the HTML used for checkboxes without
effecting the rest of the FormHelper.

markstory commented:
@ndm2 Updated now.

markstory committed:
Add missing doc block.

ADmad commented:
Build failures were unrelated.
